src/main/java/com/mentoring/ecommerce/adapter/in/web/ProductController.java

```diff
@@ -33,34 +37,46 @@
     private final ProductMapper productMapper;
 
     @PostMapping()
-    @ResponseStatus(HttpStatus.CREATED)
-    public void saveProduct(@Valid @RequestBody final ProductRequest request) {
-        saveUseCase.saveProduct(productMapper.toDomain(request));
+    @ResponseStatus(CREATED)
+    public ProductResponse saveProduct(@Valid @RequestBody final ProductRequest request) {
+        Product product = saveUseCase.saveProduct(productMapper.toDomain(request));
+        return productMapper.toResponse(product)
+                .add(linkTo(methodOn(ProductController.class).findProductById(product.getId())).withSelfRel())
+                .add(linkTo(methodOn(ProductController.class).findAllProducts()).withRel("products"));
     }
 
     @GetMapping
-    public ResponseEntity<List<ProductResponse>> findAllProducts() {
-        return ResponseEntity.ok().body(findUseCase.findAll().stream()
+    public List<ProductResponse> findAllProducts() {
+        return findUseCase.findAll().stream()
                 .map(productMapper::toResponse)
-                .toList());
+                .map(product -> product.add(linkTo(
+                        methodOn(ProductController.class).findProductById(product.getId())).withSelfRel()))
+                .collect(Collectors.toList());
     }
 
     @GetMapping("{productId}")
-    public ResponseEntity<ProductResponse> findProductById(@PathVariable(name = "productId") final Integer id) {
-        return ResponseEntity.ok().body(productMapper.toResponse(findUseCase.findById(id)));
+    public ProductResponse findProductById(@PathVariable(name = "productId") final Integer id) {
+        return productMapper.toResponse(findUseCase.findById(id))
+                .add(linkTo(ProductController.class).slash(id).withSelfRel())
+                .add(linkTo(methodOn(ProductController.class).findProductById(id)).withRel("update"))
```

Comment:
I couldn't find a non-complex way to get the `update` method to add to the link because of the 1st argument type (ProductRequest).

I made it works through `mapper` manipulation but didn't like the result, because doesn't seem right to mess with the `MapperImpl` (figure).

This brings me two questions:

- How to get the method when a parameter is required? Demands specific manipulation?
- How do I deal with `mappers` differents than `toResponse` and `toDomain` when using `MapStruct`, should be done as said here: [(item 3.3, page 15, Adding custom methods to mappers)](https://mapstruct.org/documentation/dev/reference/pdf/mapstruct-reference-guide.pdf)?


![image](https://user-images.githubusercontent.com/68197280/162191058-d92362c9-6c59-4ac3-8c9a-4d3a4669f1bc.png)

Reply:
That is a great question! I just realized that in my code I have no such scenario. I can think of some things:

- Passing a null body
- Creating a 'dummy' example body 

But honestly neither seems right to me. I searched a bit about this and found no straight answers. I asked this on StackOverflow, let's wait to see if someone in the community has a nice solution: https://stackoverflow.com/questions/71912090/representing-a-request-body-on-hate-oas-link